core/src/main/java/org/mule/runtime/core/DefaultMessageExecutionContext.java

```diff
@@ -6,58 +6,110 @@
  */
 package org.mule.runtime.core;
 
+import static java.time.OffsetTime.now;
 import static java.util.Optional.ofNullable;
 
+import java.io.Serializable;
+import java.time.OffsetTime;
+import java.util.Optional;
+
 import org.mule.runtime.core.api.MessageExecutionContext;
+import org.mule.runtime.core.api.MuleContext;
 import org.mule.runtime.core.api.MuleEvent;
+import org.mule.runtime.core.api.config.MuleConfiguration;
+import org.mule.runtime.core.api.construct.FlowConstruct;
 import org.mule.runtime.core.api.source.MessageSource;
 
-import java.io.Serializable;
-import java.util.Date;
-import java.util.Optional;
-
 /**
  * Default immutable implementation of {@link MessageExecutionContext}.
  *
  * @since 4.0
  */
-public class DefaultMessageExecutionContext implements MessageExecutionContext, Serializable {
+public final class DefaultMessageExecutionContext implements MessageExecutionContext, Serializable {
 
   private static final long serialVersionUID = -3664490832964509653L;
 
+  /**
+   * Builds a new execution context with the given parameters.
+   * 
+   * @param context the {@link MuleContext} of the application that is building this context.
+   * @param flow the flow that processes events of this context. 
+   * @param correlationId See {@link MessageExecutionContext#getCorrelationId()}.
+   */
+  public static MessageExecutionContext buildContext(MuleContext context, FlowConstruct flow) {
```

Comment:
FlowConstruct has MuleContext already doesn't it?  I think `create` works better here as this isn't a builder.